pkg/apis/config/v1alpha1/types.go

```diff
@@ -0,0 +1,276 @@
+// Copyright The NRI Plugins Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/containers/nri-plugins/pkg/apis/config/v1alpha1/instrumentation"
+	"github.com/containers/nri-plugins/pkg/apis/config/v1alpha1/log"
+	"github.com/containers/nri-plugins/pkg/apis/config/v1alpha1/resmgr/control"
+	"github.com/containers/nri-plugins/pkg/apis/config/v1alpha1/resmgr/policy/balloons"
+	"github.com/containers/nri-plugins/pkg/apis/config/v1alpha1/resmgr/policy/template"
+	"github.com/containers/nri-plugins/pkg/apis/config/v1alpha1/resmgr/policy/topologyaware"
+)
+
+// TopologyAwarePolicy represents the configuration for the topology-aware policy.
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +genclient
+type TopologyAwarePolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Policy topologyaware.Config `json:"policy"`
+	// +optional
+	Common CommonConfig `json:"common,omitempty"`
```

Comment:
Two comments/questions:

1. Maybe I'm just too stuck with the common spec vs. status pattern in the APIs but the immediate question rose that why not follow it? 😇 Spec specifies the desired state (i.e. complete configuration) and status the tells where we are in getting there. WDYT?

2. Another note is that the `common` field just looks strange from the users point of view (imho). Common what(?) Policy as a separate field makes more sense. I'd suggest putting the common stuff just as fields on the top level

```golang
type TopologyAwarePolicy struct {
	metav1.TypeMeta
	metav1.ObjectMeta

	Policy topologyaware.Config
	// Common stuff for the resource manager
	Control control.Config
	Log log.Config
	Instrumentation instrumentation.Config
}
```

Nothing prevents having the same getter func/method (`CommonConfig()`) as before for the common config bits, but it doesn't need to be visible in the user-facing API.

With these changes the the CR yaml would look smth like:

```yaml
apiVersion: config.nri/v1alpha1
kind: TopologyAwarePolicy
metadata:
  name: default
spec:
  policy:
    pinCPU: true
    pinMemory: true
    reservedResources:
      cpu: 750m
  log:
    klog:
      skip_headers: true
  instrumentation:
    reportPeriod: 60s
```

Thoughts?

Reply:
That (providing a uniform access API via common mandatory interfaces as opposed to a common mandatory struct type) is a very good point. I completely missed/overlooked that possibility. I need to think about it and probably give it a try.

Reply:
> 1. Maybe I'm just too stuck with the common spec vs. status pattern in the APIs but the immediate question rose that why not follow it? 😇 Spec specifies the desired state (i.e. complete configuration) and status the tells where we are in getting there. WDYT?

Uhmmm, isn't that what we have here ? At least that's what I was aiming for. The /status subresource describes the status of applying the configuration described by the resource for all nodes that it is applicable to. No ?

Reply:
> Uhmmm, isn't that what we have here ?

Not really, all the "desired state" stuff are on the main level, same as status, i.e. we currently have like

```yaml
apiVersion: config.nri/v1alpha1
kind: TopologyAwarePolicy
metadata:
  name: default
policy:
  pinCPU: true
  ...
log:
  ...
status:
 ...
```

What I was suggesting have a separate `spec` field on the main level, under which all the "desired state" stuff would go, like


```yaml
apiVersion: config.nri/v1alpha1
kind: TopologyAwarePolicy
metadata:
  name: default
spec:
  policy:
    pinCPU: true
    ...
  log:
    ...
status:
 ...
```

Reply:
Oh, I forgot about another possibility is to make the common part "inline", similar to e.g. ObjectMeta, e.g.
```suggestion
	Policy topologyaware.Config `json:"policy"`
	// +optional
	Common CommonConfig `json:",inline"
```

Optionally maybe renaming the type to `CommonPolicyConfig` or `CommonResourcePolicyConfig`, with a look forward to some common config patterns of other type of plugins...